PR: Update Variable Explorer page for Spyder 4 (#108)
spyder-ide/spyder-docs: doc/variableexplorer.rst

Variable Explorer
#################

> The **Variable Explorer** allows you to view and manage interactively the variables previously generated in the Editor or the IPython Console. 

Reviewer: ```suggestion
The **Variable Explorer** allows you to view and manage interactively the variables previously generated after code execution in the IPython Console. 
```

.. image:: images/variable_explorer/variable-explorer-standard.png
   :align: center
   :alt: Spyder Variable Explorer, with a list of variables and their contents

|

> It shows the namespace contents (all global object references, such as variables, functions, modules, etc.) of the currently selected :doc:`ipythonconsole` session, and allows you to modify their value, add more values or delete them through a variety of GUI-based editors.

Reviewer: ```suggestion
It shows the namespace contents (all global object references, such as variables, class instances, etc.) of the currently selected :doc:`ipythonconsole` session, and allows you to modify their value, add more values or delete them through a variety of GUI-based editors.
```

.. image:: images/variable_explorer/variable-explorer-execution.gif
   :align: center
   :alt: Spyder Variable Explorer execution with a variable of type list

|

The Variable Explorer gives you information on the name, size, type and value of each one of the variables. To modify the value of one of them, you can double click it in the pane and write its new value.

.. image:: images/variable_explorer/variable-explorer-modifying.gif
   :align: center
   :alt: Spyder Variable Explorer modifying value of a variable 

|

==============
Object Viewers
==============

> Spyder's :guilabel:`Variable Explorer` offers built in support for editing lists, strings, dictionaries, NumPy arrays, Pandas DataFrames, and more, and can also histogram, plot, or even display some of them as an RGB image. Several examples of this functionality are shown below:

Reviewer: ```suggestion
Spyder's :guilabel:`Variable Explorer` offers built in support for editing lists, strings, dictionaries, NumPy arrays, Pandas DataFrames and Series, and more, and can also plot them, or even display some as an RGB image. Several examples of this functionality are shown below:
```

|

~~~~~~~
Strings
~~~~~~~

When a string variable is longer than forty characters, you can double click it to see its value in a text editor where you can modify it.

.. image:: images/variable_explorer/variable-explorer-text-long.png
   :align: center
   :alt: Variable Explorer text editor, displaying a long string in a window

|

~~~~~~~~~~~~
Dictionaries
~~~~~~~~~~~~

> When the type of a variable is a dictionary, double clicking it will show a viewer displaying each one of the keys of the dictionary along with its associated value. You can double click any of the values to modify it. 

Reviewer 2: ```suggestion
Double-clicking on dictionaries will show a viewer displaying each of its keys with its associated value. You can double click any of the values to modify them, which will open a new viewer if the value is itself an object.
```

> When the type of a variable is a dictionary, double clicking it will show you a viewer displaying each one of the keys of the dictionary along with its associated value. You can double click any of the values to modify it. 

Reviewer: ```suggestion
When the type of a variable is a dictionary, double clicking it will show a viewer displaying each one of the keys of the dictionary along with its associated value. You can double click any of the values to modify it. 
```

.. image:: images/variable_explorer/variable-explorer-dictionary.png
   :align: center
   :alt: Dictionary editor displaying keys and their types, sizes, and values

|

~~~~~
Lists
~~~~~

> For variables that are of type list, the Variable Explorer only displays the first ten values of the list. If you want to see them all, double click the variable to open a viewer that will display the index, type, size and value of each element of the list. You can also modify any value by double clicking it in the viewer and writing the new one.

Reviewer 2: ```suggestion
For lists, the main Variable Explorer displays a preview of the first ten values. To see them all, double click the list to open a viewer that will display the index, type, size and value of each element of the list. Just like dictionaries, you can double-click values to edit them.

> For variables that are of type list, the window pane only displays the first ten values of the list. If you want to see them all, double click the variable in the pane to open a viewer that will display the index, type, size and value of each element of the list. You can also modify any value by double clicking it in the viewer and writing the new one.

Reviewer: ```suggestion
For variables that are of type list, the Variable Explorer only displays the first ten values of its values. If you want to see them all, double click the variable to open a viewer that will display the index, type, size and value of each element of the list. You can also modify any value by double clicking it in the viewer and writing the new one.
```

Author: I would stick to "the Variable Explorer only displays the first ten values of the list." Otherwise, the word values will be repeated.

Reviewer: Done.

.. image:: images/variable_explorer/variable-explorer-list.png
   :align: center
   :alt: List editor displaying a list, showing one being edited

|

> 

Reviewer 2: ```suggestion

```

~~~~~~~~~~~~
Numpy Arrays
~~~~~~~~~~~~

> For Numpy Arrays, the window pane only shows some of the values of the array. You can open a viewer that shows you graphically in a "heat map" the values of the array, displaying each value in a colored box, depending on the size of the value. You can deactivate the background color by unchecking the option in the window. If the array is too large, background color will be automatically deactivated.

Reviewer: ```suggestion
For Numpy Arrays, the Variable Explorer only shows some of their values. By double-clicking on them, a viewer will be shown that displays graphically the array values in a "heat map", i.e. it displays each value in a colored box, depending on the size of the value. You can deactivate the background color by unchecking the option in the viewer. If the array is too large, the background color will be automatically deactivated.
```

Author: I think it will be better to say "By double-clicking them, a viewer will be shown displaying graphically the array values in a "heat map" but @CAM-Gerlach, what do you think?

Author: I think it will be better to say "By double-clicking them, a viewer will be shown graphically displaying the array values in a "heat map" but @CAM-Gerlach, what do you think?

Reviewer 2: By double-clicking them, a viewer will be shown graphically displaying the array values in a "heat map

Reviewer: Ok, I changed the text to what you suggested.

> For Numpy Arrays, the Variable Explorer only shows some of their values. By double-clicking them, a viewer will be shown graphically displaying the array values in a "heat map", i.e. it displays each value in a colored box, depending on the size of the value. You can deactivate the background color by unchecking the option in the viewer. If the array is too large, the background color will be automatically deactivated.

Reviewer 2: ```suggestion
Like lists, for Numpy arrays the Variable Explorer shows a preview of their values. Double-clicking them will open a viewer displaying the array values in a "heat map", with each value in a grid cell colored based on its numeric quantity. You can deactivate the background color by unchecking the appropriate option in the viewer, which will happen automatically if the array is too large to improve performance.
```

> .. image:: images/variable_explorer/variable-explorer-array-2D-resize.png

Reviewer: The values of this image should have more differences between each other is more clearly visible the heatmap you are talking about.

Reviewer 2: Absolutely.

   :align: center
   :alt: Array editor with a 2D int array, displaying a "heatmap" of its values

|

> If possible, you can also change the format of the values in the array, choosing the number of decimals that you want the array to display. For this, click the option format and in the window displayed you can set the desired format, using printf formatting. For more information go to "https://docs.python.org/3/library/stdtypes.html#printf-style-bytes-formatting".

Reviewer: ```suggestion
If possible, you can also change the format of the values in the array, choosing the number of decimals that you want the array to display. For this, click the option format and in the new window that appears you can set the desired format, using ``printf`` formatting. For more information about this, please go to "https://docs.python.org/3/library/stdtypes.html#printf-style-bytes-formatting".
```

Reviewer: Please also use a rest link to reference the webpage you're adding in this paragraph.

> If possible, you can also change the format of the values in the array, choosing the number of decimals that you want the array to display. For this, click the option format and in the new window that appears you can set the desired format, using ``printf`` formatting. For more information about this, please go to "https://docs.python.org/3/library/stdtypes.html#printf-style-bytes-formatting".

Reviewer 2: ```suggestion
If supported by the datatype, you can also change the format of the array's values, choosing the number of decimals that you want the array to display. For this, click the :guilabel:`Format` button and and set the desired formatting in the dialog that appears, using standard `Printf-style syntax`_.

.. _Printf-style syntax: https://docs.python.org/3/library/stdtypes.html#printf-style-bytes-formatting
```

> 

Reviewer 2: ```suggestion
```

> Additionally, you can adjust the size of the rows and columns of the array by moving the arrows. Clicking the option resize, will bring the chart back to its default size.

Reviewer: ```suggestion
Additionally, you can adjust the size of the rows and columns of the array by expanding or contracting the column headers. Clicking the option Resize will bring the chart back to its default size.
```

> Additionally, you can adjust the size of the rows and columns of the array by expanding or contracting the column headers. Clicking the option Resize will bring the chart back to its default size.

Reviewer 2: ```suggestion
Additionally, you can adjust the size of the rows and columns of the array by expanding or contracting their headers. Clicking the :guilabel:`Resize` will set it automatically.
```

.. image:: images/variable_explorer/variable-explorer-array-resize.gif
   :align: center
   :alt: Array editor with a 2D int array, showing resizing of columns
 

~~~~~~~~~~

> DataFrames

Reviewer: The Dataframes section should be before the Numpy Arrays one because they are more important.

Reviewer 2: On the basis of what? Dataframes are more specialized than arrays, and particularly in many disciplines arrays are much more commonly used than dataframes if the latter are even used at all. Furthermore, arrays are simpler and form the basis of dataframes, so they should naturally be discussed first.

Author: I agree with CAM on this one.

Reviewer: Ok, no problem. It's really not critical and it'll avoid @juanis2112 to do some rewriting when moving these sections around too.

> 

Reviewer 2: ```suggestion
|

```

~~~~~~~~~~

> For DataFrames, same as Numpy Arrays, you can see their values in a viewer where you can deactivate background color, change the format and resize the rows and columns of the table in which information is displayed. 

Reviewer 2: ```suggestion
DataFrames, like Numpy arrays, display in a viewer where you can show or hide "heatmap" colors, change the format and resize the rows and columns either manually or automatically.
```

.. image:: images/variable_explorer/variable-explorer-dataframe.png
   :align: center

>    :alt: Dataframe editor showing data frame "heatmap"

Reviewer: We should also include an image for multi-indexed dataframes, such as the one shown in [our blog](https://www.spyder-ide.org/blog/spyder-variable-explorer/).

Reviewer 2: Agreed.

Author: Done

Reviewer: Great!

> |

Reviewer: At this point please add a section about the Object Explorer, which allows to view the internals of any Python object (again, as described in [our blog](https://www.spyder-ide.org/blog/spyder-variable-explorer/)).

Reviewer 2: Yes.

Author: Done

> 

Reviewer 2: ```suggestion

```

============
Options Menu
============

> The options menu on the top left of the Variable Explorer allows you to control what you are able to see in the Variable Explorer. 

Reviewer: ```suggestion
The options menu on the top right of the Variable Explorer allows you to control what you are able to see in it. 
```

> The options menu on the top right of the Variable Explorer allows you to control what you are able to see in it. 

Reviewer 2: ```suggestion
The options menu in the top right of the Variable Explorer pane allows you filter the variables shown using several criteria. 
```

.. image:: images/variable_explorer/variable-explorer-menu.png
   :align: center
   :alt: Spyder Variable Explorer, with options menu

|

> It also allows you to show the min and max values of a Numpy array in the window pane, instead of its values.

Reviewer: ```suggestion
It also allows you to show the min and max values of Numpy arrays, instead of its values.
```

> It also allows you to show the min and max values of Numpy arrays, instead of its values.

Reviewer 2: ```suggestion
It also allows you to display the min and max of Numpy arrays instead of a preview of their values.
```

.. image:: images/variable_explorer/variable-explorer-array-min&max.png
   :align: center
   :alt: Variable Explorer showing max and min values of numpy array

|

> 

Reviewer 2: @juanis2112 Still needs to be applied.

===============
Toolbar Options
===============

> On top of the variable explorer, you can find the toolbar options which allow other interactions with the variables in the explorer. You can save data in the Variable Explorer as a file with spydata extension, which can be later imported to recover all the variables stored.

Reviewer: ```suggestion
On top of the Variable Explorer, you can find the toolbar options, which allow other interactions with the variables in the explorer. For example, you can save the data in the Variable Explorer as a file with spydata extension, which can be later imported to recover all the variables stored.
```

> On top of the Variable Explorer, you can find the toolbar options, which allow other interactions with the variables in the explorer. For example, you can save the data in the Variable Explorer as a file with spydata extension, which can be later imported to recover all the variables stored.

Reviewer 2: ```suggestion
The Variable Explorer's toolbar includes several useful features that affect the entire namespace. For example, you can save the current session's data as a `.spydata` file, which can be loaded later to recover all the variables stored.
```

.. image:: images/variable_explorer/variable-explorer-import-data.gif
   :align: center
   :alt: Variable Explorer showing how to save and import data

|

> You will also find the option to remove all the variables from the variable explorer and the option to search variables by its name or type, filtering according to the text entered in the search box.

Reviewer: ```suggestion
You will also find the option to remove all variables from the Variable Explorer and the option to search variables by its name or type, filtering according to the text entered in the search box.
```

> You will also find the option to remove all variables from the Variable Explorer and the option to search variables by its name or type, filtering according to the text entered in the search box.

Reviewer 2: ```suggestion
There is also a button to remove all displayed variables, and a search box to find objects by  name or type.
```

.. image:: images/variable_explorer/variable-explorer-search.gif
   :align: center
   :alt: Variable Explorer showing how to search variables

|

> Finally you will find an option to refresh the value of the variables in the Variable Explorer while running code in the Editor.

Reviewer: ```suggestion
Finally you will find an option to refresh the Variable Explorer while running code in the IPython console.
```

> Finally you will find an option to refresh the Variable Explorer while running code in the IPython console.

Reviewer 2: ```suggestion
Finally, there is a button to refresh the Variable Explorer's contents, which will update it to show the current state of the code running in the IPython console.
```

(MISSING GIF)

> 

Reviewer 2: ```suggestion

```

================

> Advanced Options

Reviewer 2: ```suggestion
Advanced Functionality
```

================

> You can also get an additional menu of options by right clicking on top of any of the variables. Through this menu, you can rename, remove or edit any variable. You can also make a copy of it by clicking the option "duplicate", which will open a dialog box in which you can write the name of the new copy of this variable.

Reviewer 2: ```suggestion
The context menu, available by right-clicking any variable, provides numerous additional options to interact with objects of various types. These include renaming, removing or editing existing variables, as well as the :guilabel:`duplicate` option to create a new copy one of under a new name you enter in the resulting dialog box.
```

> You can also get another menu of options by right clicking on top of any of the variables on the window pane. In this menu, you can rename, remove or edit any variable in the Variable Explorer. You can also make a copy of it by clicking the option "duplicate" which will open a dialog box in which you can write the name of the new copy of this variable.

Reviewer: ```suggestion
You can also get an additional menu of options by right clicking on top of any of the variables. Through this menu, you can rename, remove or edit any variable. You can also make a copy of it by clicking the option "duplicate", which will open a dialog box in which you can write the name of the new copy of this variable.
```

.. image:: images/variable_explorer/variable-explorer-duplicate.gif
   :align: center
   :alt: Variable Explorer showing duplicating a variable

| 

> Besides, you can copy and paste the value of a variable, saving it in the Variable Explorer with any name that you choose. This allows you to change the type of the variable that you are pasting which turns out to be really useful, for example if you want to copy the elements of a list into an array.

Reviewer 2: ```suggestion
Furthermore, you can copy and paste the value of a variable, saving it in the Variable Explorer with any name that you choose. This allows you to change the type of the variable that you are pasting which can be very useful, allowing to to, for example, easily copy the elements of a list into an array.
```

> You can also copy and paste the value of a variable saving it in the Variable Explorer with any name that you choose. This allows you to change the type of the variable that you are pasting which turns out to be really useful, for example if you want to copy the elements of a list into an array.

Reviewer: ```suggestion
Besides, you can copy and paste the value of a variable, saving it in the Variable Explorer with any name that you choose. This allows you to change the type of the variable that you are pasting which turns out to be really useful, for example if you want to copy the elements of a list into an array.
```

.. image:: images/variable_explorer/variable-explorer-copy-paste.gif
   :align: center
   :alt: Variable Explorer showing copying list into array

| 

> Additionally, you can create a variable from zero directly in the Variable Explorer by clicking the option Insert, which allows you to write the key (which should be in quotation marks) and the value for the variable that you want to insert.

Reviewer 2: ```suggestion
Additionally, you can create a variable from scratch directly in the Variable Explorer with the :guilabel:`Insert` option, which allows you to type the key (which should be in quotation marks) and the value for the item that you want to insert. In addition to adding a new top-level variable, this feature also allows you to create a new key in a dictionary, a new element in a list, and much more.
```

> However, you can also create a variable from cero directly in the Variable Explorer by clicking the option insert which will allow you to write the key (which should be between quotation marks) and the value for the variable that you want to insert.

Reviewer: ```suggestion
Additionally, you can create a variable from zero directly in the Variable Explorer by clicking the option Insert, which allows you to write the key (which should be in quotation marks) and the value for the variable that you want to insert.
```

.. image:: images/variable_explorer/variable-explorer-insert.gif
   :align: center
   :alt: Variable Explorer showing insertion of a new variable

| 

> For lists and Numpy arrays, you will find more advanced options which include plotting or making a histogram using the data of the variable.

Reviewer 2: ```suggestion
For lists and Numpy arrays, more advanced options are available, including generating plots and histograms of their values appropriate to their type and dimensions.
```

> For lists and numpy arrays, you will find more advanced options which include plotting or making a histogram using the data of the variable.

Reviewer: ```suggestion
For lists and Numpy arrays, you will find more advanced options which include plotting or making a histogram using the data of the variable.
```

.. image:: images/variable_explorer/variable-explorer-histogram.png
   :align: center
   :alt: Plot window showing a histogram, generated via the previous options

> |

Reviewer 2: Avoid a double image where practicable.

.. image:: images/variable_explorer/variable-explorer-plot.png
   :align: center
   :alt: Plot window showing a plot, generated via the previous options

|

> You can save an array in disk too, in a file with npy extension, by clicking the option on this menu. 

Reviewer 2: ```suggestion
You can even save an array to a ``.npy`` file by simply clicking the appropriate option, which can later be loaded by Spyder or in your code via `numpy.load()`.
```

> You can also save an array in disk in a file with npy extension, by clicking the option in this menu. 

Reviewer: ```suggestion
You can save an array in disk too, in a file with npy extension, by clicking the option on this menu. 
```

.. image:: images/variable_explorer/variable-explorer-contextmenu-array.png
   :align: center
   :alt: Context menu for an int array, with the Show image option selected

|

> In the case of two-dimensional arrays, you can also show an image representing pixels using the numbers of the array as RGB values. For this, Spyder uses Matplotlib's colormaps, which can be easily changed according to your own preferences. To learn how to do this, you can go to "https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html"

Reviewer 2: ```suggestion
For two-dimensional arrays, you can also display them as images, treating their values as RGB colors. For this, Spyder uses Matplotlib's colormaps, which can be `easily changed to match your preferences`_.

.. _easily changed to match your preferences: https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html
```

If its using the values as RGB, doesn't it need to be three dimensional? It sounds like its just grayscale if its colored via Matplotlib's colormaps, not RGB.

> In the case of multidimensional arrays, you can also show an image representing pixels using the numbers of the array as RGB values. For this, Spyder uses Matplotlib's colormaps, which can be easily changed according to your own preferences. To learn how to do this, you can go to "https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html"

Reviewer: ```suggestion
In the case of two-dimensional arrays, you can also show an image representing pixels using the numbers of the array as RGB values. For this, Spyder uses Matplotlib's colormaps, which can be easily changed according to your own preferences. To learn how to do this, you can go to "https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html"
```

Reviewer: Please use a rest link to replace the webpage you are mentioning here.

.. image:: images/plot-window/plot-window-show-image.png
   :align: center
   :alt: Plot window showing an interactive image based on the array's data

|

> 

Reviewer 2: ```suggestion

```

Variable Explorer
#################

The **Variable Explorer** allows you to interactively browse and manage the variables and objects generated running your code.

.. image:: images/variable_explorer/variable-explorer-execution.gif
   :width: 500
   :align: center
   :alt: Spyder Variable Explorer execution with a variable of type list

|

It shows the namespace contents (including all global objects, variables, class instances and more) of the currently selected :doc:`ipythonconsole` session, and allows you to add, remove, and edit their values through a variety of GUI-based editors.

.. image:: images/variable_explorer/variable-explorer-standard.png
   :align: center
   :alt: Spyder Variable Explorer, with a list of variables and their contents

|

The Variable Explorer gives you information on the name, size, type and value of each object. To modify a scalar variable, like an number, string or boolean, simply double click it in the pane and type its new value.

> .. image:: images/variable_explorer/variable-explorer-modifying.gif

Reviewer: This one shows a visual bug in Spyder: the spinner never stops. I think the problem is that the signal to tell it stop is never emitted when editing inline objects. I'll fix it on the Spyder side.

Author: Done

Reviewer 2: LGTM

Reviewer: Thanks @juanis2112 for fixing the bug!

   :align: center
   :alt: Spyder Variable Explorer modifying value of a variable 

|

==============
Object Viewers
==============

Spyder's :guilabel:`Variable Explorer` offers built in support for editing lists, strings, dictionaries, NumPy arrays, Pandas DataFrames, Series and more; as well as being able to plot and visualize them with one click.

~~~~~~~
Strings
~~~~~~~

When a string variable is longer than forty characters, you can double click it to see its value in a text editor to more easily modify it.

> .. image:: images/variable_explorer/variable-explorer-text-long.png

Reviewer: This one has an error in the first line: "Recieve" must be "Receive". Also, I would only like to see the first line (i.e. not showing from `Parameters` down) to not be confused with the functionality of showing docstrings of the Object Explorer.

Reviewer 2: But that defeats the point of the long text editor as its only one line, and furthermore is the exact opposite of what we both agreed to tell @juanis2112 to change on this screenshot that resulted in her redoing it to have more, and more "realistic", text. If we're going to change it, we need to come up with some other text to show here that is a similar size.

Reviewer: Ok, but then we shouldn't show a docstring but another kind of text. Could you come up with a different text then?

Author: I fixed the one that was already there, if you come up with a better one let me know.

Reviewer 2: I think its fine, but if @ccordoba12 insists on having something different, here's a random snippit from a CSV, that a user might conceivably read in as a string and want to parse later. Feel free to simplify as desired.

```
time,runtime,ping,adc_vb_f,adc_va_f,adc_vl_f,adc_ic_f,adc_il_f,t_hs,t_batt,t_amb,t_rts,charge_state,array_fault,vb_f,vb_ref,ahc_r,ahc_t,kwhc,load_state,load_fault,v_lvd,ahl_r,ahl_t,hourmeter,alarm,dip_switch,led_state,power_out,sweep_vmp,sweep_pmax,sweep_voc,vb_min_daily,vb_max_daily,ahc_daily,ahl_daily,array_fault_daily,load_fault_daily,alarm_daily,vb_min,vb_max,sequence_count,timestamp,crc_errors,valid_packets,bytes_read,bytes_written,bytes_remaining,packets_sent,packets_dropped
2019-11-13 00:00:29.797264,490800.0,0,12.509,0.656,12.485,0.0,1.164,0.0,1.0,1.0,25.0,3,0000000000000000,12.506,0.0,175.7,175.7,2.4,1,0000000000000000,11.484,145.8,145.8,7413,00000000000000000000000000000001,0000000000001101,6,0.0,8.435,7.172,11.249,12.469,12.527,0.0,0.6,0000000000000000,0000000000000000,00000000000000000000000000000001,9.439,19.47,231814,2019-11-13 14:36:23.389000,0,867,49871585280,19074114444,236913627136,0,0
2019-11-13 00:01:29.797262,490860.0,0,12.5,0.656,12.463,0.0,1.169,0.0,1.0,1.0,25.0,3,0000000000000000,12.503,0.0,175.7,175.7,2.4,1,0000000000000000,11.484,145.8,145.8,7413,00000000000000000000000000000001,0000000000001101,6,0.0,8.435,7.172,11.249,12.469,12.527,0.0,0.6,0000000000000000,0000000000000000,00000000000000000000000000000001,9.439,19.47,231874,2019-11-13 14:37:23.396000,0,867,49879449600,19074114444,236913627136,0,0
2019-11-13 00:02:29.797257,490920.0,0,12.497,0.656,12.463,0.0,1.167,0.0,1.0,1.0,25.0,3,0000000000000000,12.5,0.0,175.7,175.7,2.4,1,0000000000000000,11.484,145.8,145.8,7413,00000000000000000000000000000001,0000000000001101,6,0.0,8.435,7.172,11.249,12.469,12.527,0.0,0.6,0000000000000000,0000000000000000,00000000000000000000000000000001,9.439,19.47,231934,2019-11-13 14:38:23.405000,0,867,49887576064,19074114444,236913627136,0,0
2019-11-13 00:03:29.797264,490980.0,0,12.491,0.656,12.46,0.0,1.193,0.0,1.0,1.0,25.0,3,0000000000000000,12.494,0.0,175.7,175.7,2.4,1,0000000000000000,11.484,145.8,145.8,7413,00000000000000000000000000000001,0000000000001101,6,0.0,8.435,7.172,11.249,12.469,12.527,0.0,0.6,0000000000000000,0000000000000000,00000000000000000000000000000001,9.439,19.47,231994,2019-11-13 14:39:23.414000,0,867,49895440384,19074114444,236913627136,0,0
```

Reviewer: Let's leave it as it is.

| 

Furthermore, you can copy and paste the value of a variable, saving it in the Variable Explorer with any name that you choose. This allows you to change the type of the variable that you are pasting which can be very useful, allowing to, for example, easily copy the elements of a list into an array.

.. image:: images/variable_explorer/variable-explorer-copy-paste.gif
   :align: center
   :alt: Variable Explorer showing copying list into array

| 

Additionally, you can create a variable from scratch directly in the Variable Explorer with the :guilabel:`Insert` option, which allows you to type the key (which should be in quotation marks) and the value for the item that you want to insert. In addition to adding a new top-level variable, this feature also allows you to create a new key in a dictionary, a new element in a list, and much more.

.. image:: images/variable_explorer/variable-explorer-insert.gif
   :align: center
   :alt: Variable Explorer showing insertion of a new variable

| 

For lists and Numpy arrays, more advanced options are available, including generating plots and histograms of their values appropriate to their type and dimensions.

> .. image:: images/variable_explorer/variable-explorer-histogram-plot.gif

Reviewer: This is a bit confusing because you have to move from the Variable Explorer to Plots to show the generated images, Instead, I suggest to hide the Editor, show the Variable Explorer on the left half of the entire window and Plots on the right half. That wayt it'd be easier for users to understand what's going on.

Author: Done

Reviewer 2: LGTM

   :align: center
   :alt: Plot window showing a plot, generated via the previous options

|

You can even save an array to a ``.npy`` file by simply clicking the appropriate option, which can later be loaded by Spyder or in your code via `numpy.load()`.

.. image:: images/variable_explorer/variable-explorer-contextmenu-array.png
   :width: 500
   :align: center
   :alt: Context menu for an int array, with the Show image option selected

|

For two-dimensional arrays, you can also display them as images, treating their values as RGB colors. For this, Spyder uses Matplotlib's colormaps, which can be `easily changed to match your preferences`_.

.. _easily changed to match your preferences: https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html

> .. image:: images/variable_explorer/variable-explorer-show-image.gif

Reviewer: Same here: I suggest to show only the Variable Explorer (left) and Plots (right) on the gif.

Author: Done

Reviewer 2: LGTM
